eva_sub_cli/semantic_metadata.py

```diff
@@ -0,0 +1,117 @@
+import yaml
+
+from retry import retry
+from ebi_eva_common_pyutils.ena_utils import download_xml_from_ena
+from ebi_eva_common_pyutils.logger import AppLogger
+
+
+PROJECT_KEY = 'project'
+ANALYSIS_KEY = 'analysis'
+SAMPLE_KEY = 'sample'
+FILES_KEY = 'files'
+PARENT_PROJECT_KEY = 'parentProject'
+CHILD_PROJECTS_KEY = 'childProjects'
+PEER_PROJECTS_KEY = 'peerProjects'
+BIOSAMPLE_OBJECT_KEY = 'bioSampleObject'
+CHARACTERISTICS_KEY = 'characteristics'
+TAX_ID_KEY = 'taxId'
+ANALYSIS_ALIAS_KEY = 'analysisAlias'
+
+
+def cast_list(l, type_to_cast=str):
+    for e in l:
+        yield type_to_cast(e)
+
+
+class SemanticMetadataChecker(AppLogger):
+
+    def __init__(self, metadata):
+        self.metadata = metadata
+        self.errors = []
+
+    def write_result_yaml(self, output_path):
+        with open(output_path, 'w') as open_yaml:
+            yaml.safe_dump(data=self.errors, stream=open_yaml)
+
+    def check_all(self):
+        self.check_all_project_accessions()
+        self.check_all_taxonomy_codes()
+        self.check_analysis_alias_coherence()
+
+    def check_all_project_accessions(self):
+        """Check that ENA project accessions exist and are public."""
+        project = self.metadata[PROJECT_KEY]
+        if PARENT_PROJECT_KEY in project:
+            self.check_project_accession(project[PARENT_PROJECT_KEY], f'/{PROJECT_KEY}/{PARENT_PROJECT_KEY}')
+        for idx, accession in enumerate(project.get(CHILD_PROJECTS_KEY, [])):
+            self.check_project_accession(accession, f'/{PROJECT_KEY}/{CHILD_PROJECTS_KEY}/{idx}')
+        for idx, accession in enumerate(project.get(PEER_PROJECTS_KEY, [])):
+            self.check_project_accession(accession, f'/{PROJECT_KEY}/{PEER_PROJECTS_KEY}/{idx}')
+
+    def check_all_taxonomy_codes(self):
+        """Check that taxonomy IDs are valid according to ENA."""
+        project = self.metadata[PROJECT_KEY]
+        self.check_taxonomy_code(project[TAX_ID_KEY], f'/{PROJECT_KEY}/{TAX_ID_KEY}')
+        # Check sample taxonomies for novel samples
+        for idx, sample in enumerate(self.metadata[SAMPLE_KEY]):
+            if BIOSAMPLE_OBJECT_KEY in sample:
+                self.check_taxonomy_code(sample[BIOSAMPLE_OBJECT_KEY][CHARACTERISTICS_KEY][TAX_ID_KEY][0]['text'],
+                                         f'/{SAMPLE_KEY}/{idx}/{BIOSAMPLE_OBJECT_KEY}/{CHARACTERISTICS_KEY}/{TAX_ID_KEY}')
+
+    @retry(tries=4, delay=2, backoff=1.2, jitter=(1, 3))
+    def check_project_accession(self, project_acc, json_path):
+        try:
+            download_xml_from_ena(f'https://www.ebi.ac.uk/ena/browser/api/xml/{project_acc}')
+        except Exception:
+            self.add_error(json_path, f'{project_acc} does not exist or is private')
+
+    @retry(tries=4, delay=2, backoff=1.2, jitter=(1, 3))
+    def check_taxonomy_code(self, taxonomy_code, json_path):
+        try:
+            download_xml_from_ena(f'https://www.ebi.ac.uk/ena/browser/api/xml/{taxonomy_code}')
```

Comment:
I think somewhere in this function we should cache the call to ENA so that if there are 1000 sample  with the same taxonomy we only make one call.

Reply:
Good point, will do